### PR TITLE
Optimize put - Dont try to fetch filecache for not existing filecache…

### PR DESCRIPTION
**Pull request by mrow4a**
Gets rid of one useless query:
selection_111

### CODE REVIEW

File: lib/private/Files/Storage/Wrapper/Encryption.php [hunk 1]
```diff
@@ -909,7 +909,8 @@ protected function parseRawHeader($rawHeader) {
 	 */
 	protected function getHeader($path) {
 		$realFile = $this->util->stripPartialFileExtension($path);
-		if ($this->storage->file_exists($realFile)) {
+		$exists = $this->storage->file_exists($realFile);
+		if ($exists) {
 			$path = $realFile;
```
Comment (mrow4a) [inline lib/private/Files/Storage/Wrapper/Encryption.php]: if `realFile` exists, we substitude $path with its real path

File: lib/private/Files/Storage/Wrapper/Encryption.php [hunk 2]
```diff
@@ -921,8 +922,9 @@ protected function getHeader($path) {
 		if (!isset($result[Util::HEADER_ENCRYPTION_MODULE_KEY])) {
 			if (!empty($result)) {
 				$result[Util::HEADER_ENCRYPTION_MODULE_KEY] = 'OC_DEFAULT_MODULE';
-			} else {
+			} else if ($exists) {
 				// if the header was empty we have to check first if it is a encrypted file at all
+				// We would do query to filecache only if we know that entry in filecache exists
 				$info = $this->getCache()->get($path);
```
Comment (mrow4a) [inline lib/private/Files/Storage/Wrapper/Encryption.php]: Though, if real file does not exists, it make no sense to fetch filecache which does not exists and we know it earlier..

**Comment by PVince81** [comment]
Nice catch!

File: lib/private/Files/Storage/Wrapper/Encryption.php [hunk 3]
```diff
@@ -921,8 +922,9 @@ protected function getHeader($path) {
...
 				if (isset($info['encrypted']) && $info['encrypted'] === true) {
 					$result[Util::HEADER_ENCRYPTION_MODULE_KEY] = 'OC_DEFAULT_MODULE';
```
Comment (mrow4a) [inline lib/private/Files/Storage/Wrapper/Encryption.php]: If filecache does not exists, this line will never be executed

**Comment by mrow4a** [comment]
Please mind this is going not to master but to branch optimize_propfind. I will try to batch all small optimisations. Or should I do that directly to master?

File: lib/private/Files/Storage/Wrapper/Encryption.php (hunk 2 above)
Comment (jvillafanez) [inline lib/private/Files/Storage/Wrapper/Encryption.php]: Is it possible that the file exists but it isn't cached? What could happend in this case?

**Comment by jvillafanez** [comment]
I'm just asking: Does it make sense to get the header of a partial file? Why the behaviour for part files is different depending on whether the "real" file exists or not in the backend?

File: lib/private/Files/Storage/Wrapper/Encryption.php (hunk 2 above)
Comment (mrow4a) [inline lib/private/Files/Storage/Wrapper/Encryption.php]: Real file will not exist without filecache entry. Otherwise is a bug

**Comment by mrow4a** [comment]
To my understanding part file is just temporary file which then is moved, and it does not have any filecache so we should avoid doing unnecessary calls

File: lib/private/Files/Storage/Wrapper/Encryption.php (hunk 2 above)
Comment (jvillafanez) [inline lib/private/Files/Storage/Wrapper/Encryption.php]: Nobody expects the spanish inquisition!!
Maybe the file is uploaded directly in the backend and ownCloud doesn't have any metadata. A direct propfind request to the target file could trigger this.

We shouldn't expect that anyone will follow the workflow we want.

> If file exists and you call it, you get false in return and wasted 1 DB call.

If that's the worst case I think it's fine.

**Comment by jvillafanez** [comment]
👍 

I think we should review how the part files are being handled as a whole, at least for the encryption. Anyway, that's outside of this PR's scope.

File: lib/private/Files/Storage/Wrapper/Encryption.php (hunk 2 above)
Comment (PVince81) [inline lib/private/Files/Storage/Wrapper/Encryption.php]: I think getHeader is anyway only called at the beginning, not after uploading a file directly to the backend

**Comment by PVince81** [comment]
note: there are integration tests on master with encryption enabled, so need to keep an eye on those.

**Pull request merged**